Add FilterByQuery and ExecuteQuery as simple operators receiving textual python query

dafnapension opened the pull request:
It seems that the simple conditions, listed in `FilterByCondition` , can all be replaced by simple lambda conditions.  We can thus replace, and enjoy from additional, perhaps more involved, conditions that lambda can express. 
This also refers to [an issue](https://github.com/IBM/unitxt/issues/415) by @matanor

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -1475,7 +1532,7 @@ class AddFieldNamePrefix(StreamInstanceOperator):
     """Adds a prefix to each field name in each instance of a stream.
 
     Args:
-        prefix_dict (Dict[str, str]): A dictionary mapping stream names to prefixes.
```
Operator `AddFieldNamePrefix` is used nowhere. Shall I remove it?

elronbandel replied:
yes looks unnecessary

dafnapension replied:
Removed.. as part of this branch. If this branch will not be merged into main, I will remember to remove `AddFieldNamePrefix` in another PR..

elronbandel reviewed (changes requested):
The direction is really promising. 
I think a better machinism will be to use eval directly:
```
>>> eval("x < y", None, {"x":5, "y":7})
True
>>> eval("x < y", None, {"x":5, "y":3})
False
```
That way we can have complicated conditions consisted of many variables. 

so the final api can be:
`operator = FilterByQuery(query="len(target) > 5 or target == 'yes'")`

Accordingly we can add also general operator:
`operator = ExecuteQuery(query="target + '_end'",  to_field="target")`

dafnapension commented:
> The direction is really promising. I think a better machinism will be to use eval directly:
> 
> ```
> >>> eval("x < y", None, {"x":5, "y":7})
> True
> >>> eval("x < y", None, {"x":5, "y":3})
> False
> ```
> 
> That way we can have complicated conditions consisted of many variables.
> 
> so the final api can be: `operator = FilterByQuery(query="len(target) > 5 or target == 'yes'")`
> 
> Accordingly we can add also a general operator: `operator = ExecuteQuery(query="target + '_end'", to_field="target")`

A neat idea!!  replaced FilterByLambdaCondition to FilterByQuery, and implemented.  Indeed neat!

dafnapension commented:
@elronbandel , this is really neat! I now added `ExecuteQuery` per your idea! I think it may be handy for metrics (I am now digging into issue 439)

yoavkatz commented:
I also encountered this problem that may pose a new requirement.

We have for each instance two fields with llists
   labels = ['person', 'location', 'person']
   entity = ['Jane,     'New York',  'John']
   
   and I want to filter both lists together to and keep only 'persons'.
   
   labels = ['person',  'person']
   entity = ['Jane',  'John']
   
   How could this be handled?

dafnapension commented:
Hi, @yoavkatz ,  per [your question](https://github.com/IBM/unitxt/pull/449#issuecomment-1881205717), I added a unittest that does just that.
```
operator = ExecuteQuery(
            queries_to_fields=[
                ['[e for l,e in zip(labels, entity) if l=="person"]', "entity"],
                ['[l for l in labels if l == "person"]', "labels"],
            ]
        )
```

matanor commented:
> The direction is really promising. I think a better machinism will be to use eval directly:
> 
> ```
> >>> eval("x < y", None, {"x":5, "y":7})
> True
> >>> eval("x < y", None, {"x":5, "y":3})
> False
> ```
> 
> That way we can have complicated conditions consisted of many variables.
> 
> so the final api can be: `operator = FilterByQuery(query="len(target) > 5 or target == 'yes'")`
> 
> Accordingly we can add also a general operator: `operator = ExecuteQuery(query="target + '_end'", to_field="target")`

Suggestion: what i was originally thinking of is this:

`operator = Filter(lambda x: len(x[target]) > 5 or x[target] == 'yes')`

Or

`Execute(lambda x: x[target] + '_end', to_field="target")`

I think this avoids the need to express python code in strings. This of for example having a syntax error in the code string.. would be easier IMO if its just a lambda expression.

yoavkatz commented:
A few points:
1. I agree that adding a lambda expression is better than string. Although at the end, we will store in the json the string representation, and need to maintain this option.
2. Why support multiple queries in same operator call?  I think it will be clearer  to have it one by one, as in Matan's example: Execute(lambda x: x["target"] + '_end', to_field="target")

dafnapension commented:
I thought that in the spirit of `FieldOperator` , we may want to complete a mission in one call, and a series of executions can do it easier than squashing somehow to one execution. Like with your example of labels and entities.

yoavkatz commented:
> I thought that in the spirit of `FieldOperator` , we may want to complete a mission in one call, and a series of executions can do it easier than squashing somehow to one execution. Like with your example of labels and entities.

You know by take, that this is complicated API.  I think we should discuss with Elron.

elronbandel commented on src/unitxt/operators.py:
```diff
@@ -1263,6 +1263,87 @@ def _is_required(self, instance: dict) -> bool:
+
+    """
+
+    queries: List[str]
```
```suggestion
    query: str # "x > 0 and y < 3"
```

elronbandel commented on src/unitxt/operators.py:
```diff
@@ -1263,6 +1263,87 @@ def _is_required(self, instance: dict) -> bool:
+
+    """
+
+    queries_to_fields: List[List[str]]
```
```suggestion
    query: str
    to_field: str
```

dafnapension commented:
> > I thought that in the spirit of `FieldOperator` , we may want to complete a mission in one call, and a series of executions can do it easier than squashing somehow to one execution. Like with your example of labels and entities.
> 
> You know by take, that this is complicated API. I think we should discuss with Elron.

@elronbandel agrees with you, @yoavkatz , so I replaced the list of queries by a single query.

elronbandel reviewed (approved):
Amazing work!

dafnapension commented:
Thank you, @elronbandel and @yoavkatz  for suggesting the operators, and for your reviews and improvements

yoavkatz commented:
Before we release it to uses I suggest a few additional changes (for a new PR)

Rename "FilterByQuery" to "FilterByExpression".
Rename "ExecuteQuery" to "ExecuteExpression".
